### PR TITLE
Assorted refactor

### PR DESCRIPTION
**Pull request by jasl**
Dentaku is an awesome gem!

I'm working on a prototype app for researching that can define a form with many fields (it support many datatypes) and give a expression to evaluate value.

the demo is in <https://dentaku-calculator.herokuapp.com/> 
source code in <https://github.com/jasl/dentaku_calculator>
Dentaku relates is in <https://github.com/jasl/dentaku_calculator/blob/master/app/lib/virtual_model.rb#L8-L34>

define a form with fields
image

fill fields and expression
image

integration with model validation (support i18n)
image
image

calculate
image

It's based on Dentaku, but I do some refactor to support my needs, the major change (but not break compatibility) is classified errors and stored some meta to helping top-level given user a friendly tips.

So I give this PR for discussion, is these changes good to Dentaku? how can I polish these changes? English isn't my mother tongue, the reason of errors follows practice of English?

### CODE REVIEW

**Comment by cweyer** [comment]
👍 
We do similar validations on formulas for [Keyline](https://www.keyline-mis.com) and currently just show generic information ("parse error"), would be way nicer to be able to actually show where the parse error occurred, so +1 from me for this PR

File: lib/dentaku/ast/arithmetic.rb [hunk 1]
```diff
@@ -65,13 +71,15 @@ def validate_value(val)
 
       def validate_operation(val)
         unless val.respond_to?(operator)
-          fail Dentaku::ArgumentError, "#{ self.class } requires operands that respond to #{ operator }"
+          raise Dentaku::ArgumentError.for(:not_respond_operator, operation: self.class, operator: operator),
```
Comment (rubysolo) [inline lib/dentaku/ast/arithmetic.rb]: I think `:invalid_operator` would read better than `:not_respond_operator`.

File: spec/exceptions_spec.rb [hunk 2]
```diff
@@ -1,9 +0,0 @@
-require 'spec_helper'
-require 'dentaku/exceptions'
-
-describe Dentaku::UnboundVariableError do
-  it 'includes variable name(s) in message' do
-    exception = described_class.new(['length'])
-    expect(exception.message).to match /length/
```
Comment (rubysolo) [inline spec/exceptions_spec.rb]: Do we still test that the `UnboundVariableError` message includes the variable name(s)?

**Comment by jasl** [comment]
fixed

**Comment by rubysolo** [comment]
🎉 Thank you!

**Pull request merged**